kratos/mpi/utilities/pointer_communicator.h

```diff
@@ -0,0 +1,373 @@
+
+
+#if !defined(KRATOS_POINTER_COMMUNICATOR_H_INCLUDED )
+#define  KRATOS_POINTER_COMMUNICATOR_H_INCLUDED
+
+
+// System includes
+#include <string>
+#include <iostream>
+
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+#include "includes/data_communicator.h"
+#include "includes/global_pointer.h"
+#include "includes/mpi_serializer.h"
+#include "containers/global_pointers_vector.h"
+#include "containers/global_pointers_unordered_map.h"
+
+#include "mpi/utilities/mpi_coloring_utilities.h"
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+template< class TPointerDataType, class TSendType >
+class ResultsProxy
+{
+public:
+
+    ResultsProxy(
+        int current_rank,
+        GlobalPointersUnorderedMap< TPointerDataType, TSendType > NonLocalData,
+        std::function< TSendType(GlobalPointer<TPointerDataType>&) > user_function
+    ):
+        mcurrent_rank(current_rank), mNonLocalData(NonLocalData), muser_function(user_function)
+    {}
+
+    /// Destructor.
+    virtual ~ResultsProxy() {}
+
+    /**this function returns the effect of "user_function(gp)" both if the gp is locally owned
+     * and if it is remotely owned
+     */
+    TSendType Get(GlobalPointer<TPointerDataType>& gp)
+    {
+        if(gp.GetRank() == mcurrent_rank)
+            return muser_function(gp);
+        else
+            return mNonLocalData[gp];
+    }
+
+private:
+    const int mcurrent_rank;
+    GlobalPointersUnorderedMap< TPointerDataType, TSendType > mNonLocalData;
+    std::function< TSendType(GlobalPointer<TPointerDataType>&) > muser_function;
+
+};
+
+
+/// Short class definition.
+/** Detail class definition.
+*/
+template< class TPointerDataType >
+class GlobalPointerCommunicator
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of GlobalPointerCommunicator
+    KRATOS_CLASS_POINTER_DEFINITION(GlobalPointerCommunicator);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    GlobalPointerCommunicator(DataCommunicator& rComm, GlobalPointersVector< TPointerDataType >& gp_list ):
+        mrDataCommunicator(rComm)
+    {
+        AddPointers(gp_list.begin(),gp_list.end());
+    }
+
+    template< class TIteratorType >
+    GlobalPointerCommunicator(DataCommunicator& rComm, TIteratorType begin, TIteratorType end):
+        mrDataCommunicator(rComm)
+    {
+        AddPointers(begin,end);
+    }
+
+    /// Destructor.
+    virtual ~GlobalPointerCommunicator() {}
+
+    /**this function applies the input function onto the remote data
+     * and retrieves the result to the caller rank
+     */
+    template< class TSendType >
+    ResultsProxy<TPointerDataType, TSendType> Apply(std::function< TSendType(GlobalPointer<TPointerDataType>&) > user_function)
```

Comment:
Can/should the argument of the function to be applied remotely be const (`const GlobalPointer<TPointerDataType>&` or maybe `GlobalPointer<const TPointerDataType>&`?). In either case, I assume that we do not want to allow the global pointer operation to modify the remote object.

Reply:
why not? imagine you want to mark a flag as visited...